Model/PageType.php

```diff
@@ -0,0 +1,223 @@
+    {
+        if ($this->pageType === null) {
+            $this->pageType = $this->resolvePageInfo();
+        }
+
+        return $this->pageType;
+    }
+
+    public function getJson(): string
+    {
+        $pageType = $this->getInfo();
+
+        return $pageType ? $this->json->serialize($pageType) : '';
+    }
+
+    public function getPageType(): ?string
+    {
+        $pageInfo = $this->getInfo();
+
+        return $pageInfo['type'] ?? null;
+    }
+
+    public function setContext(Context $context): self
+    {
+        $this->context = $context;
+
+        return $this;
+    }
+
+    public function getContext(): ?Context
+    {
+        return $this->context;
+    }
+
+    public function resolvePageInfo(): ?array
+    {
+        $storeId = (int) $this->storeManager->getStore()->getId();
+        $request = $this->getContext()->get('request')->toArray();
+        $urlParts = $request['url'];
+        $url = $urlParts['pathname'];
+        $result = null;
+
+        if (substr($url, 0, 1) === '/' && $url !== '/') {
+            $url = ltrim($url, '/');
+        }
+        $this->redirectType = 0;
+        $customUrl = $this->customUrlLocator->locateUrl($url);
+        $url = $customUrl ?: $url;
+        $finalUrlRewrite = $this->findFinalUrl($url, $storeId);
+        if ($finalUrlRewrite) {
+            $relativeUrl = $finalUrlRewrite->getRequestPath();
+            $resultArray = $this->rewriteCustomUrls($finalUrlRewrite, $storeId) ?? [
+                    'id' => $finalUrlRewrite->getEntityId(),
+                    'canonical_url' => $relativeUrl,
+                    'relative_url' => $relativeUrl,
+                    'redirectCode' => $this->redirectType,
+                    'redirect_code' => $this->redirectType,
+                    'type' => $this->sanitizeType($finalUrlRewrite->getEntityType())
+                ];
+            if (!empty($urlParts['search'])) {
+                $resultArray['relative_url'] .= $urlParts['search'];
+            }
+
+            if (empty($resultArray['id'])) {
+                throw new \RuntimeException(
+                    __('No such entity found with matching URL key: %url', ['url' => $url])
+                );
+            }
+
+            $result = $resultArray ?: null;
+        }
+
+        return $result;
+    }
+
+
+    /**
+     * Handle custom urls with and without redirects
+     *
+     * @param UrlRewrite $finalUrlRewrite
+     * @param int $storeId
+     * @return array|null
+     */
+    private function rewriteCustomUrls(UrlRewrite $finalUrlRewrite, int $storeId): ?array
+    {
+        if ($finalUrlRewrite->getEntityType() === 'custom' || !($finalUrlRewrite->getEntityId() > 0)) {
+            $finalCustomUrlRewrite = clone $finalUrlRewrite;
+            $finalUrlRewrite = $this->findFinalUrl($finalCustomUrlRewrite->getTargetPath(), $storeId, true);
+            $relativeUrl =
+                $finalCustomUrlRewrite->getRedirectType() == 0
```

Comment:
Please consider using more strict '===' here (hidden types casting will not be applied anymore).

Reply:
Code from core, but fixed in this module